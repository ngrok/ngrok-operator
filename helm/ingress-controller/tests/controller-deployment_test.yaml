# yaml-language-server: $schema=https://raw.githubusercontent.com/quintush/helm-unittest/master/schema/helm-testsuite.json
suite: test controller-deployment
templates:
- controller-deployment.yaml
# The following included templates are needed due to the way helm unittest works.
# It won't load the templates unless they are included here. Due to the checksums
# we are including on the deployment, we need to include them here. It makes the
# snapshot much larger than it needs to be, but it's the only way to test the
# deployment at this time.
- controller-rbac.yaml
- rbac/role.yaml
tests:
- it: Should match default snapshot
  asserts:
  - matchSnapshot: {}
- it: Should match all-options snapshot
  set:
    extraEnv:
      TEST_ENV_VAR: test
      SECRET_ENV_VAR:
        secretKeyRef:
          key: test-key
        value: test-value
    extraVolumes:
    - name: test-volume
      emptyDir: {}
    extraVolumeMounts:
    - name: test-volume
      mountPath: /test-volume
  asserts:
  - matchSnapshot: {}
- it: Should use the specified secret name for the credentials secret
  set:
    credentials.secret.name: test-secret-name
  template: controller-deployment.yaml
  documentIndex: 0 # Document 0 is the deployment since its the first template
  asserts:
  - matchRegex:
      path: spec.template.spec.containers[0].env[0].valueFrom.secretKeyRef.name
      pattern: test-secret-name
- it: Should pass the region via container args to the controller if specified
  set:
    region: eu
  template: controller-deployment.yaml
  documentIndex: 0 # Document 0 is the deployment since its the first template
  asserts:
  - matchRegex:
      path: spec.template.spec.containers[0].args[0]
      pattern: --region=eu
- it: Should not pass the region via container args to the controller if not specified
  template: controller-deployment.yaml
  documentIndex: 0 # Document 0 is the deployment since its the first template
  asserts:
  - notMatchRegex:
      path: spec.template.spec.containers[0].args[0]
      pattern: --region=eu
- it: Should pass extra metadata to the deployment
  set:
    metaData:
      metaDataKey1: metaDataValue1
      metaDataKey2: metaDataValue2
  template: controller-deployment.yaml
  documentIndex: 0 # Document 0 is the deployment since its the first template
  asserts:
  - matchRegex:
      path: spec.template.spec.containers[0].args[0]
      pattern: --metadata=metaDataKey1=metaDataValue1,metaDataKey2=metaDataValue2
- it: Should pass through extra volumes and extra volume mounts
  set:
    extraVolumes:
    - name: test-volume
      emptyDir: {}
    extraVolumeMounts:
    - name: test-volume
      mountPath: /test-volume
  template: controller-deployment.yaml
  documentIndex: 0 # Document 0 is the deployment since its the first template
  asserts:
  - matchRegex:
      path: spec.template.spec.volumes[0].name
      pattern: test-volume
  - matchRegex:
      path: spec.template.spec.containers[0].volumeMounts[0].name
      pattern: test-volume
  - matchRegex:
      path: spec.template.spec.containers[0].volumeMounts[0].mountPath
      pattern: /test-volume
- it: Should pass log level argument if set
  set:
    log:
      level: error
  template: controller-deployment.yaml
  documentIndex: 0 # Document 0 is the deployment since its the first template
  asserts:
  - contains:
      path: spec.template.spec.containers[0].args
      content: --zap-log-level=error
- it: Should pass log format argument if set
  set:
    log:
      format: console
  template: controller-deployment.yaml
  documentIndex: 0 # Document 0 is the deployment since its the first template
  asserts:
  - contains:
      path: spec.template.spec.containers[0].args
      content: --zap-encoder=console
- it: Should pass stacktrace log level argument if set
  set:
    log:
      stacktraceLevel: error
  template: controller-deployment.yaml
  documentIndex: 0 # Document 0 is the deployment since its the first template
  asserts:
  - contains:
      path: spec.template.spec.containers[0].args
      content: --zap-stacktrace-level=error
- it: Defaults to having "soft" pod anti-affinity
  template: controller-deployment.yaml
  documentIndex: 0 # Document 0 is the deployment since its the first template
  asserts:
  - equal:
      path: spec.template.spec.affinity.podAntiAffinity
      value:
        preferredDuringSchedulingIgnoredDuringExecution:
        - podAffinityTerm:
            labelSelector:
              matchLabels:
                app.kubernetes.io/component: controller
                app.kubernetes.io/instance: RELEASE-NAME
                app.kubernetes.io/name: kubernetes-ingress-controller
            topologyKey: kubernetes.io/hostname
          weight: 1
- it: Easily allows for setting a "hard" pod anti-affinity
  template: controller-deployment.yaml
  documentIndex: 0 # Document 0 is the deployment since its the first template
  set:
    podAntiAffinityPreset: "hard"
  asserts:
  - equal:
      path: spec.template.spec.affinity.podAntiAffinity
      value:
        requiredDuringSchedulingIgnoredDuringExecution:
        - labelSelector:
            matchLabels:
              app.kubernetes.io/component: controller
              app.kubernetes.io/instance: RELEASE-NAME
              app.kubernetes.io/name: kubernetes-ingress-controller
          topologyKey: kubernetes.io/hostname
- it: Allows overriding the full affinity
  template: controller-deployment.yaml
  documentIndex: 0 # Document 0 is the deployment since its the first template
  set:
    affinity:
      nodeAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
          nodeSelectorTerms:
          - matchExpressions:
            - key: topology.kubernetes.io/zone
              operator: In
              values:
              - antarctica-east1
              - antarctica-west1
        preferredDuringSchedulingIgnoredDuringExecution:
        - weight: 1
          preference:
            matchExpressions:
            - key: another-node-label-key
              operator: In
              values:
              - another-node-label-value
  asserts:
  - equal:
      path: spec.template.spec.affinity
      value:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
            - matchExpressions:
              - key: topology.kubernetes.io/zone
                operator: In
                values:
                - antarctica-east1
                - antarctica-west1
          preferredDuringSchedulingIgnoredDuringExecution:
          - weight: 1
            preference:
              matchExpressions:
              - key: another-node-label-key
                operator: In
                values:
                - another-node-label-value
- it: Defaults to not specifying a priority class
  template: controller-deployment.yaml
  documentIndex: 0 # Document 0 is the deployment since its the first template
  asserts:
  - isNull:
      path: spec.template.spec.priorityClassName
- it: Allows setting a priority class
  template: controller-deployment.yaml
  documentIndex: 0 # Document 0 is the deployment since its the first template
  set:
    priorityClassName: "high-priority"
  asserts:
  - equal:
      path: spec.template.spec.priorityClassName
      value: high-priority
