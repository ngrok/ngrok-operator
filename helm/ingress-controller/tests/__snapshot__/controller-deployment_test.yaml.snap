Should match all-options snapshot:
  1: |
    apiVersion: apps/v1
    kind: Deployment
    metadata:
      annotations:
        checksum/controller-role: e82dcab3ad4d03739f251c7b07a2af02c401d664726314ec52eefc966f8d7c60
        checksum/rbac: d31fdcb337a6f1ee71323040c2cbc4d5580d73ae5f7623cd19be57db97f748c1
      labels:
        app.kubernetes.io/component: controller
        app.kubernetes.io/instance: RELEASE-NAME
        app.kubernetes.io/managed-by: Helm
        app.kubernetes.io/name: kubernetes-ingress-controller
        app.kubernetes.io/part-of: kubernetes-ingress-controller
        app.kubernetes.io/version: 0.4.0
        helm.sh/chart: kubernetes-ingress-controller-0.6.0
      name: RELEASE-NAME-kubernetes-ingress-controller-manager
      namespace: NAMESPACE
    spec:
      replicas: 1
      selector:
        matchLabels:
          app.kubernetes.io/component: controller
          app.kubernetes.io/instance: RELEASE-NAME
          app.kubernetes.io/name: kubernetes-ingress-controller
      template:
        metadata:
          annotations:
            kubectl.kubernetes.io/default-container: manager
            prometheus.io/path: /metrics
            prometheus.io/port: "8080"
            prometheus.io/scrape: "true"
          labels:
            app.kubernetes.io/component: controller
            app.kubernetes.io/instance: RELEASE-NAME
            app.kubernetes.io/name: kubernetes-ingress-controller
        spec:
          containers:
          - args:
            - --controller-name=k8s.ngrok.com/ingress-controller
            - --health-probe-bind-address=:8081
            - --metrics-bind-address=:8080
            - --election-id=RELEASE-NAME-kubernetes-ingress-controller-leader
            command:
            - /manager
            env:
            - name: NGROK_API_KEY
              valueFrom:
                secretKeyRef:
                  key: API_KEY
                  name: RELEASE-NAME-kubernetes-ingress-controller-credentials
            - name: NGROK_AUTHTOKEN
              valueFrom:
                secretKeyRef:
                  key: AUTHTOKEN
                  name: RELEASE-NAME-kubernetes-ingress-controller-credentials
            - name: POD_NAMESPACE
              valueFrom:
                fieldRef:
                  fieldPath: metadata.namespace
            - name: SECRET_ENV_VAR
              value:
                secretKeyRef:
                  key: test-key
                value: test-value
            - name: TEST_ENV_VAR
              value: test
            image: docker.io/ngrok/kubernetes-ingress-controller:latest
            imagePullPolicy: IfNotPresent
            livenessProbe:
              httpGet:
                path: /healthz
                port: 8081
              initialDelaySeconds: 15
              periodSeconds: 20
            name: ngrok-ingress-controller
            readinessProbe:
              httpGet:
                path: /readyz
                port: 8081
              initialDelaySeconds: 5
              periodSeconds: 10
            resources:
              limits: {}
              requests: {}
            securityContext:
              allowPrivilegeEscalation: false
            volumeMounts:
            - mountPath: /test-volume
              name: test-volume
          serviceAccountName: RELEASE-NAME-kubernetes-ingress-controller
          volumes:
          - emptyDir: {}
            name: test-volume
  2: |
    apiVersion: rbac.authorization.k8s.io/v1
    kind: Role
    metadata:
      name: ngrok-ingress-controller-leader-election-role
      namespace: NAMESPACE
    rules:
    - apiGroups:
      - ""
      resources:
      - configmaps
      verbs:
      - get
      - list
      - watch
      - create
      - update
      - patch
      - delete
    - apiGroups:
      - coordination.k8s.io
      resources:
      - leases
      verbs:
      - get
      - list
      - watch
      - create
      - update
      - patch
      - delete
    - apiGroups:
      - ""
      resources:
      - events
      verbs:
      - create
      - patch
  3: |
    apiVersion: rbac.authorization.k8s.io/v1
    kind: ClusterRole
    metadata:
      name: ngrok-ingress-controller-proxy-role
    rules:
    - apiGroups:
      - authentication.k8s.io
      resources:
      - tokenreviews
      verbs:
      - create
    - apiGroups:
      - authorization.k8s.io
      resources:
      - subjectaccessreviews
      verbs:
      - create
  4: |
    apiVersion: rbac.authorization.k8s.io/v1
    kind: RoleBinding
    metadata:
      name: ngrok-ingress-controller-leader-election-rolebinding
      namespace: NAMESPACE
    roleRef:
      apiGroup: rbac.authorization.k8s.io
      kind: Role
      name: ngrok-ingress-controller-leader-election-role
    subjects:
    - kind: ServiceAccount
      name: RELEASE-NAME-kubernetes-ingress-controller
      namespace: NAMESPACE
  5: |
    apiVersion: rbac.authorization.k8s.io/v1
    kind: ClusterRoleBinding
    metadata:
      name: ngrok-ingress-controller-manager-rolebinding
    roleRef:
      apiGroup: rbac.authorization.k8s.io
      kind: ClusterRole
      name: ngrok-ingress-controller-manager-role
    subjects:
    - kind: ServiceAccount
      name: RELEASE-NAME-kubernetes-ingress-controller
      namespace: NAMESPACE
  6: |
    apiVersion: rbac.authorization.k8s.io/v1
    kind: ClusterRoleBinding
    metadata:
      name: ngrok-ingress-controller-proxy-rolebinding
    roleRef:
      apiGroup: rbac.authorization.k8s.io
      kind: ClusterRole
      name: ngrok-ingress-controller-proxy-role
    subjects:
    - kind: ServiceAccount
      name: RELEASE-NAME-kubernetes-ingress-controller
      namespace: NAMESPACE
  7: |
    apiVersion: rbac.authorization.k8s.io/v1
    kind: ClusterRole
    metadata:
      creationTimestamp: null
      name: ngrok-ingress-controller-manager-role
    rules:
    - apiGroups:
      - ""
      resources:
      - configmaps
      verbs:
      - create
      - delete
      - get
      - list
      - update
      - watch
    - apiGroups:
      - ""
      resources:
      - events
      verbs:
      - create
      - patch
    - apiGroups:
      - ""
      resources:
      - secrets
      verbs:
      - get
      - list
      - watch
    - apiGroups:
      - ingress.k8s.ngrok.com
      resources:
      - domains
      verbs:
      - create
      - delete
      - get
      - list
      - patch
      - update
      - watch
    - apiGroups:
      - ingress.k8s.ngrok.com
      resources:
      - domains/finalizers
      verbs:
      - update
    - apiGroups:
      - ingress.k8s.ngrok.com
      resources:
      - domains/status
      verbs:
      - get
      - patch
      - update
    - apiGroups:
      - ingress.k8s.ngrok.com
      resources:
      - httpsedges
      verbs:
      - create
      - delete
      - get
      - list
      - patch
      - update
      - watch
    - apiGroups:
      - ingress.k8s.ngrok.com
      resources:
      - httpsedges/finalizers
      verbs:
      - update
    - apiGroups:
      - ingress.k8s.ngrok.com
      resources:
      - httpsedges/status
      verbs:
      - get
      - patch
      - update
    - apiGroups:
      - ingress.k8s.ngrok.com
      resources:
      - ippolicies
      verbs:
      - create
      - delete
      - get
      - list
      - patch
      - update
      - watch
    - apiGroups:
      - ingress.k8s.ngrok.com
      resources:
      - ippolicies/finalizers
      verbs:
      - update
    - apiGroups:
      - ingress.k8s.ngrok.com
      resources:
      - ippolicies/status
      verbs:
      - get
      - patch
      - update
    - apiGroups:
      - ingress.k8s.ngrok.com
      resources:
      - tcpedges
      verbs:
      - create
      - delete
      - get
      - list
      - patch
      - update
      - watch
    - apiGroups:
      - ingress.k8s.ngrok.com
      resources:
      - tcpedges/finalizers
      verbs:
      - update
    - apiGroups:
      - ingress.k8s.ngrok.com
      resources:
      - tcpedges/status
      verbs:
      - get
      - patch
      - update
    - apiGroups:
      - ingress.k8s.ngrok.com
      resources:
      - tunnels
      verbs:
      - create
      - delete
      - get
      - list
      - patch
      - update
      - watch
    - apiGroups:
      - ingress.k8s.ngrok.com
      resources:
      - tunnels/finalizers
      verbs:
      - update
    - apiGroups:
      - ingress.k8s.ngrok.com
      resources:
      - tunnels/status
      verbs:
      - get
      - patch
      - update
    - apiGroups:
      - networking.k8s.io
      resources:
      - ingressclasses
      verbs:
      - get
      - list
      - watch
    - apiGroups:
      - networking.k8s.io
      resources:
      - ingresses
      verbs:
      - get
      - list
      - update
      - watch
    - apiGroups:
      - networking.k8s.io
      resources:
      - ingresses/status
      verbs:
      - get
      - list
      - update
      - watch
Should match default snapshot:
  1: |
    apiVersion: apps/v1
    kind: Deployment
    metadata:
      annotations:
        checksum/controller-role: e82dcab3ad4d03739f251c7b07a2af02c401d664726314ec52eefc966f8d7c60
        checksum/rbac: d31fdcb337a6f1ee71323040c2cbc4d5580d73ae5f7623cd19be57db97f748c1
      labels:
        app.kubernetes.io/component: controller
        app.kubernetes.io/instance: RELEASE-NAME
        app.kubernetes.io/managed-by: Helm
        app.kubernetes.io/name: kubernetes-ingress-controller
        app.kubernetes.io/part-of: kubernetes-ingress-controller
        app.kubernetes.io/version: 0.4.0
        helm.sh/chart: kubernetes-ingress-controller-0.6.0
      name: RELEASE-NAME-kubernetes-ingress-controller-manager
      namespace: NAMESPACE
    spec:
      replicas: 1
      selector:
        matchLabels:
          app.kubernetes.io/component: controller
          app.kubernetes.io/instance: RELEASE-NAME
          app.kubernetes.io/name: kubernetes-ingress-controller
      template:
        metadata:
          annotations:
            kubectl.kubernetes.io/default-container: manager
            prometheus.io/path: /metrics
            prometheus.io/port: "8080"
            prometheus.io/scrape: "true"
          labels:
            app.kubernetes.io/component: controller
            app.kubernetes.io/instance: RELEASE-NAME
            app.kubernetes.io/name: kubernetes-ingress-controller
        spec:
          containers:
          - args:
            - --controller-name=k8s.ngrok.com/ingress-controller
            - --health-probe-bind-address=:8081
            - --metrics-bind-address=:8080
            - --election-id=RELEASE-NAME-kubernetes-ingress-controller-leader
            command:
            - /manager
            env:
            - name: NGROK_API_KEY
              valueFrom:
                secretKeyRef:
                  key: API_KEY
                  name: RELEASE-NAME-kubernetes-ingress-controller-credentials
            - name: NGROK_AUTHTOKEN
              valueFrom:
                secretKeyRef:
                  key: AUTHTOKEN
                  name: RELEASE-NAME-kubernetes-ingress-controller-credentials
            - name: POD_NAMESPACE
              valueFrom:
                fieldRef:
                  fieldPath: metadata.namespace
            image: docker.io/ngrok/kubernetes-ingress-controller:latest
            imagePullPolicy: IfNotPresent
            livenessProbe:
              httpGet:
                path: /healthz
                port: 8081
              initialDelaySeconds: 15
              periodSeconds: 20
            name: ngrok-ingress-controller
            readinessProbe:
              httpGet:
                path: /readyz
                port: 8081
              initialDelaySeconds: 5
              periodSeconds: 10
            resources:
              limits: {}
              requests: {}
            securityContext:
              allowPrivilegeEscalation: false
          serviceAccountName: RELEASE-NAME-kubernetes-ingress-controller
  2: |
    apiVersion: rbac.authorization.k8s.io/v1
    kind: Role
    metadata:
      name: ngrok-ingress-controller-leader-election-role
      namespace: NAMESPACE
    rules:
    - apiGroups:
      - ""
      resources:
      - configmaps
      verbs:
      - get
      - list
      - watch
      - create
      - update
      - patch
      - delete
    - apiGroups:
      - coordination.k8s.io
      resources:
      - leases
      verbs:
      - get
      - list
      - watch
      - create
      - update
      - patch
      - delete
    - apiGroups:
      - ""
      resources:
      - events
      verbs:
      - create
      - patch
  3: |
    apiVersion: rbac.authorization.k8s.io/v1
    kind: ClusterRole
    metadata:
      name: ngrok-ingress-controller-proxy-role
    rules:
    - apiGroups:
      - authentication.k8s.io
      resources:
      - tokenreviews
      verbs:
      - create
    - apiGroups:
      - authorization.k8s.io
      resources:
      - subjectaccessreviews
      verbs:
      - create
  4: |
    apiVersion: rbac.authorization.k8s.io/v1
    kind: RoleBinding
    metadata:
      name: ngrok-ingress-controller-leader-election-rolebinding
      namespace: NAMESPACE
    roleRef:
      apiGroup: rbac.authorization.k8s.io
      kind: Role
      name: ngrok-ingress-controller-leader-election-role
    subjects:
    - kind: ServiceAccount
      name: RELEASE-NAME-kubernetes-ingress-controller
      namespace: NAMESPACE
  5: |
    apiVersion: rbac.authorization.k8s.io/v1
    kind: ClusterRoleBinding
    metadata:
      name: ngrok-ingress-controller-manager-rolebinding
    roleRef:
      apiGroup: rbac.authorization.k8s.io
      kind: ClusterRole
      name: ngrok-ingress-controller-manager-role
    subjects:
    - kind: ServiceAccount
      name: RELEASE-NAME-kubernetes-ingress-controller
      namespace: NAMESPACE
  6: |
    apiVersion: rbac.authorization.k8s.io/v1
    kind: ClusterRoleBinding
    metadata:
      name: ngrok-ingress-controller-proxy-rolebinding
    roleRef:
      apiGroup: rbac.authorization.k8s.io
      kind: ClusterRole
      name: ngrok-ingress-controller-proxy-role
    subjects:
    - kind: ServiceAccount
      name: RELEASE-NAME-kubernetes-ingress-controller
      namespace: NAMESPACE
  7: |
    apiVersion: rbac.authorization.k8s.io/v1
    kind: ClusterRole
    metadata:
      creationTimestamp: null
      name: ngrok-ingress-controller-manager-role
    rules:
    - apiGroups:
      - ""
      resources:
      - configmaps
      verbs:
      - create
      - delete
      - get
      - list
      - update
      - watch
    - apiGroups:
      - ""
      resources:
      - events
      verbs:
      - create
      - patch
    - apiGroups:
      - ""
      resources:
      - secrets
      verbs:
      - get
      - list
      - watch
    - apiGroups:
      - ingress.k8s.ngrok.com
      resources:
      - domains
      verbs:
      - create
      - delete
      - get
      - list
      - patch
      - update
      - watch
    - apiGroups:
      - ingress.k8s.ngrok.com
      resources:
      - domains/finalizers
      verbs:
      - update
    - apiGroups:
      - ingress.k8s.ngrok.com
      resources:
      - domains/status
      verbs:
      - get
      - patch
      - update
    - apiGroups:
      - ingress.k8s.ngrok.com
      resources:
      - httpsedges
      verbs:
      - create
      - delete
      - get
      - list
      - patch
      - update
      - watch
    - apiGroups:
      - ingress.k8s.ngrok.com
      resources:
      - httpsedges/finalizers
      verbs:
      - update
    - apiGroups:
      - ingress.k8s.ngrok.com
      resources:
      - httpsedges/status
      verbs:
      - get
      - patch
      - update
    - apiGroups:
      - ingress.k8s.ngrok.com
      resources:
      - ippolicies
      verbs:
      - create
      - delete
      - get
      - list
      - patch
      - update
      - watch
    - apiGroups:
      - ingress.k8s.ngrok.com
      resources:
      - ippolicies/finalizers
      verbs:
      - update
    - apiGroups:
      - ingress.k8s.ngrok.com
      resources:
      - ippolicies/status
      verbs:
      - get
      - patch
      - update
    - apiGroups:
      - ingress.k8s.ngrok.com
      resources:
      - tcpedges
      verbs:
      - create
      - delete
      - get
      - list
      - patch
      - update
      - watch
    - apiGroups:
      - ingress.k8s.ngrok.com
      resources:
      - tcpedges/finalizers
      verbs:
      - update
    - apiGroups:
      - ingress.k8s.ngrok.com
      resources:
      - tcpedges/status
      verbs:
      - get
      - patch
      - update
    - apiGroups:
      - ingress.k8s.ngrok.com
      resources:
      - tunnels
      verbs:
      - create
      - delete
      - get
      - list
      - patch
      - update
      - watch
    - apiGroups:
      - ingress.k8s.ngrok.com
      resources:
      - tunnels/finalizers
      verbs:
      - update
    - apiGroups:
      - ingress.k8s.ngrok.com
      resources:
      - tunnels/status
      verbs:
      - get
      - patch
      - update
    - apiGroups:
      - networking.k8s.io
      resources:
      - ingressclasses
      verbs:
      - get
      - list
      - watch
    - apiGroups:
      - networking.k8s.io
      resources:
      - ingresses
      verbs:
      - get
      - list
      - update
      - watch
    - apiGroups:
      - networking.k8s.io
      resources:
      - ingresses/status
      verbs:
      - get
      - list
      - update
      - watch
